# Federal awards

This notebook analyzes federal awards data (assistance prime transactions) downloaded from [USAspending.gov](https://www.usaspending.gov/search/?hash=966f5f0446cbe09c05fb1d16dee2885a&section=map).

The objective is to find which agencies awarded funding to congressional districts 14 and 15 and certain neighborhoods in the South Bronx.

In [1]:
# importing libraries
import pandas as pd

In [2]:
df = pd.read_csv("usaspending/Assistance_PrimeTransactions_2024-11-20_H01M42S17_1.csv")
pd.set_option('display.float_format', '{:.0f}'.format) # show full value of float
pd.set_option('display.max_colwidth', None) # show full strings

/var/folders/4f/fk_hbgy14cg47tw521m3y0xw0000gn/T/ipykernel_25527/3904420668.py:1: DtypeWarning: Columns (65,67) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv("usaspending/Assistance_PrimeTransactions_2024-11-20_H01M42S17_1.csv")


In [3]:
len(df)

138297

## Setting up our working `df` 

In [4]:
# keeping only some columns, so we don't have to work with so much
filtered_df = df[["awarding_agency_name", "federal_action_obligation",
            "total_obligated_amount", "total_outlayed_amount_for_overall_award", "action_date",
            "action_date_fiscal_year", "awarding_sub_agency_name", "program_activities_funding_this_award",
             "recipient_name", "recipient_zip_code", "prime_award_transaction_recipient_cd_original",
             "cfda_title", "assistance_type_description", "transaction_description", "business_types_description",
             "usaspending_permalink"
            ]]
len(filtered_df) # should be the same as df

138297

In [5]:
filtered_df.dtypes

awarding_agency_name                              object
federal_action_obligation                        float64
total_obligated_amount                           float64
total_outlayed_amount_for_overall_award          float64
action_date                                       object
action_date_fiscal_year                            int64
awarding_sub_agency_name                          object
program_activities_funding_this_award             object
recipient_name                                    object
recipient_zip_code                               float64
prime_award_transaction_recipient_cd_original     object
cfda_title                                        object
assistance_type_description                       object
transaction_description                           object
business_types_description                        object
usaspending_permalink                             object
dtype: object

In [6]:
# convert zip floats to int
filtered_df["recipient_zip_code"] = filtered_df["recipient_zip_code"].fillna(0).astype(int)

/var/folders/4f/fk_hbgy14cg47tw521m3y0xw0000gn/T/ipykernel_25527/2500644147.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  filtered_df["recipient_zip_code"] = filtered_df["recipient_zip_code"].fillna(0).astype(int)


In [7]:
# keeping only those tagged NY-14 or NY-15, which are our target congressional districts
filtered_df = filtered_df[filtered_df["prime_award_transaction_recipient_cd_original"].isin(["NY-14", "NY-15"])]
len(filtered_df)

4064

In [8]:
# target zip codes
zip_codes = [{
        "zip_num": 10451,
        "neighborhood": "Melrose"
    },
    {
        "zip_num": 10455,
        "neighborhood": "Longwood"
    },
    {
        "zip_num": 10459,
        "neighborhood": "Longwood"
    },
    {
        "zip_num": 10454,
        "neighborhood": "Mott Haven"
    },
    {
        "zip_num": 10456,
        "neighborhood": "Morrisania"
    },
    {
        "zip_num": 10474,
        "neighborhood": "Hunts Point"
    }]

# converting list of dicts to df
zip_codes = pd.DataFrame(zip_codes)

# converting target column to list
zip_list = list(zip_codes["zip_num"])
zip_list

[10451, 10455, 10459, 10454, 10456, 10474]

In [9]:
# keeping only those with the zip codes we need
filtered_df = filtered_df[filtered_df["recipient_zip_code"].isin(zip_list)]
len(filtered_df)

452

In [10]:
# merging zip details, to make it easier for analysis?

filtered_df = pd.merge(filtered_df, zip_codes, left_on="recipient_zip_code", right_on="zip_num", how='left')
filtered_df = filtered_df.drop(columns=["zip_num"])
len(filtered_df) # should be the same as previous len()

452

### Analysis

We downsized our 138297-items-long dataset into 452 targeted rows. Now, we begin our analysis. The objective is to find which agencies or areas are getting the most or least federal funding.

In [11]:
# sorting by agency
agency_outlayed_sum_df = filtered_df.groupby("awarding_agency_name")["total_outlayed_amount_for_overall_award"]\
.sum()\
.sort_values(ascending=False)\
.reset_index()

agency_outlayed_sum_df

awarding_agency_name  \
0       Department of Health and Human Services   
1   Department of Housing and Urban Development   
2                       Department of Education   
3                         Department of Justice   
4                Social Security Administration   
5                           Department of Labor   
6             Federal Communications Commission   
7                        Department of Commerce   
8                    Department of the Treasury   
9                 Small Business Administration   
10              National Endowment for the Arts   
11                        Department of Defense   
12              Environmental Protection Agency   
13     Institute of Museum and Library Services   
14        National Endowment for the Humanities   
15                  National Science Foundation   

    total_outlayed_amount_for_overall_award  
0                                 472075150  
1                                 253902948  
2                                   3025707  
3                                   1699186  
4                                    804621  
5                                    664273  
6                                    624662  
7                                    200000  
8                                    138706  
9                                     76484  
10                                    69475  
11                                        0  
12                                        0  
13                                        0  
14                                        0  
15                                        0

In [12]:
# counting items per agency
filtered_df["awarding_agency_name"].value_counts()

awarding_agency_name
Department of Housing and Urban Development    182
Department of Health and Human Services        100
Federal Communications Commission               56
Small Business Administration                   46
Department of Education                         32
Department of Justice                           11
National Endowment for the Arts                  8
Department of Defense                            4
Department of the Treasury                       3
National Science Foundation                      2
Department of Commerce                           2
Department of Labor                              2
Social Security Administration                   1
Institute of Museum and Library Services         1
Environmental Protection Agency                  1
National Endowment for the Humanities            1
Name: count, dtype: int64

In [13]:
# sorting by subagency
subagency_outlayed_sum_df = filtered_df.groupby(["awarding_agency_name", "awarding_sub_agency_name"])["total_outlayed_amount_for_overall_award"]\
.sum()\
.sort_values(ascending=False)\
.reset_index()

subagency_outlayed_sum_df

awarding_agency_name  \
0       Department of Health and Human Services   
1   Department of Housing and Urban Development   
2       Department of Health and Human Services   
3       Department of Health and Human Services   
4                       Department of Education   
5       Department of Health and Human Services   
6                         Department of Justice   
7   Department of Housing and Urban Development   
8                Social Security Administration   
9                           Department of Labor   
10            Federal Communications Commission   
11                       Department of Commerce   
12                   Department of the Treasury   
13                Small Business Administration   
14              National Endowment for the Arts   
15                        Department of Justice   
16                        Department of Defense   
17              Environmental Protection Agency   
18                        Department of Defense   
19     Institute of Museum and Library Services   
20        National Endowment for the Humanities   
21                  National Science Foundation   

                                         awarding_sub_agency_name  \
0                    Health Resources and Services Administration   
1   Assistant Secretary for Housing--Federal Housing Commissioner   
2                        Administration for Children and Families   
3       Substance Abuse and Mental Health Services Administration   
4                                         Department of Education   
5                      Centers for Disease Control and Prevention   
6                                      Office of Justice Programs   
7      Assistant Secretary for Community Planning and Development   
8                                  Social Security Administration   
9                          Employment and Training Administration   
10                              Federal Communications Commission   
11                National Oceanic and Atmospheric Administration   
12                                       Internal Revenue Service   
13                                  Small Business Administration   
14                                National Endowment for the Arts   
15                                  Offices, Boards and Divisions   
16                                       Defense Logistics Agency   
17                                Environmental Protection Agency   
18                                         Department of the Army   
19                       Institute of Museum and Library Services   
20                          National Endowment for the Humanities   
21                                    National Science Foundation   

    total_outlayed_amount_for_overall_award  
0                                 395487806  
1                                 252853066  
2                                  65136631  
3                                   9754853  
4                                   3025707  
5                                   1695860  
6                                   1679186  
7                                   1049882  
8                                    804621  
9                                    664273  
10                                   624662  
11                                   200000  
12                                   138706  
13                                    76484  
14                                    69475  
15                                    20000  
16                                        0  
17                                        0  
18                                        0  
19                                        0  
20                                        0  
21                                        0

In [14]:
subagency_outlayed_sum_df.to_csv("subagency_totals.csv", encoding="UTF-8", index=False)

In [15]:
# sorting by zip code 
zip_outlayed_sum_df = filtered_df.groupby(["recipient_zip_code", "neighborhood"])["total_outlayed_amount_for_overall_award"]\
.sum()\
.sort_values(ascending=False)\
.reset_index()

zip_outlayed_sum_df

recipient_zip_code neighborhood  total_outlayed_amount_for_overall_award
0               10459     Longwood                                444480285
1               10455     Longwood                                107663975
2               10454   Mott Haven                                 78745589
3               10456   Morrisania                                 58769286
4               10451      Melrose                                 40403107
5               10474  Hunts Point                                  3218969

In [16]:
zip_outlayed_sum_df.to_csv("zip_totals.csv", encoding="UTF-8", index=False)

In [17]:
# counting items per zip code
filtered_df["recipient_zip_code"].value_counts()

recipient_zip_code
10459    122
10454     80
10451     78
10456     77
10455     77
10474     18
Name: count, dtype: int64

In [18]:
# sorting by zip code and agency
zip_agency_outlayed_sum_df = filtered_df.groupby(["recipient_zip_code", "awarding_agency_name"])["total_outlayed_amount_for_overall_award"]\
.sum()\
.sort_values(ascending=False)\
.reset_index()

zip_agency_outlayed_sum_df

recipient_zip_code                         awarding_agency_name  \
0                10459      Department of Health and Human Services   
1                10455  Department of Housing and Urban Development   
2                10459  Department of Housing and Urban Development   
3                10454      Department of Health and Human Services   
4                10456  Department of Housing and Urban Development   
5                10454  Department of Housing and Urban Development   
6                10451  Department of Housing and Urban Development   
7                10456      Department of Health and Human Services   
8                10455      Department of Health and Human Services   
9                10451      Department of Health and Human Services   
10               10474  Department of Housing and Urban Development   
11               10455                      Department of Education   
12               10455                        Department of Justice   
13               10451               Social Security Administration   
14               10451                      Department of Education   
15               10454                          Department of Labor   
16               10455                          Department of Labor   
17               10451                       Department of Commerce   
18               10454            Federal Communications Commission   
19               10456            Federal Communications Commission   
20               10451                   Department of the Treasury   
21               10455            Federal Communications Commission   
22               10451            Federal Communications Commission   
23               10454              National Endowment for the Arts   
24               10451                Small Business Administration   
25               10454                        Department of Justice   
26               10455              National Endowment for the Arts   
27               10456                Small Business Administration   
28               10459                Small Business Administration   
29               10474            Federal Communications Commission   
30               10451              National Endowment for the Arts   
31               10474                Small Business Administration   
32               10454                Small Business Administration   
33               10459            Federal Communications Commission   
34               10455                Small Business Administration   
35               10456              National Endowment for the Arts   
36               10454                      Department of Education   
37               10451                  National Science Foundation   
38               10451        National Endowment for the Humanities   
39               10451     Institute of Museum and Library Services   
40               10451              Environmental Protection Agency   
41               10474                        Department of Justice   
42               10451                        Department of Justice   
43               10455                        Department of Defense   

    total_outlayed_amount_for_overall_award  
0                                 378433988  
1                                  90901620  
2                                  66027471  
3                                  46572676  
4                                  33762017  
5                                  31574391  
6                                  28447254  
7                                  24829214  
8                                  12290237  
9                                   9949035  
10                                  3190195  
11                                  2322460  
12                                  1679186  
13                                   804621  
14                                   703247  
15                                   350000  
16                        

In [19]:
zip_agency_outlayed_sum_df.to_csv("zip_agency_totals.csv", encoding="UTF-8", index=False)

In [20]:
# sorting by business type
biztype_sum_df = filtered_df.groupby("business_types_description")["total_outlayed_amount_for_overall_award"]\
.sum()\
.sort_values(ascending=False)\
.reset_index()

biztype_sum_df

business_types_description  \
0                                                                                                                                       OTHER   
1                                                                                         FOR-PROFIT ORGANIZATION (OTHER THAN SMALL BUSINESS)   
2                                                             NONPROFIT WITH 501C3 IRS STATUS (OTHER THAN AN INSTITUTION OF HIGHER EDUCATION)   
3                                                          NONPROFIT WITHOUT 501C3 IRS STATUS (OTHER THAN AN INSTITUTION OF HIGHER EDUCATION)   
4                                                                                     PUBLIC/STATE CONTROLLED INSTITUTION OF HIGHER EDUCATION   
5                                                                                                                                  INDIVIDUAL   
6                                 INDEPENDENT SCHOOL DISTRICT;NONPROFIT WITH 501C3 IRS STATUS (OTHER THAN AN INSTITUTION OF HIGHER EDUCATION)   
7                                                                                                                              SMALL BUSINESS   
8                                                                                                                 INDEPENDENT SCHOOL DISTRICT   
9                                                                                                                 CITY OR TOWNSHIP GOVERNMENT   
10                                                                                                                          COUNTY GOVERNMENT   
11  NONPROFIT WITH 501C3 IRS STATUS (OTHER THAN AN INSTITUTION OF HIGHER EDUCATION);FOR-PROFIT ORGANIZATION (OTHER THAN SMALL BUSINESS);OTHER   
12                                                      NONPROFIT WITH 501C3 IRS STATUS (OTHER THAN AN INSTITUTION OF HIGHER EDUCATION);OTHER   
13                                                                                                                           STATE GOVERNMENT   

    total_outlayed_amount_for_overall_award  
0                                 366006416  
1                                 251160019  
2                                 110893116  
3                                   3577582  
4                                   1507868  
5                                     51949  
6                                     48888  
7                                     24535  
8                                     10840  
9                                         0  
10                                        0  
11                                        0  
12                                        0  
13                                        0

In [21]:
# sorting by agency and business type
agency_biztype_sum_df = filtered_df.groupby(["awarding_agency_name", "business_types_description"])["total_outlayed_amount_for_overall_award"]\
.sum()\
.sort_values(ascending=False)\
.reset_index()

agency_biztype_sum_df

awarding_agency_name  \
0       Department of Health and Human Services   
1   Department of Housing and Urban Development   
2       Department of Health and Human Services   
3       Department of Health and Human Services   
4                       Department of Education   
5                         Department of Justice   
6   Department of Housing and Urban Development   
7   Department of Housing and Urban Development   
8                Social Security Administration   
9                       Department of Education   
10                          Department of Labor   
11            Federal Communications Commission   
12            Federal Communications Commission   
13                       Department of Commerce   
14                   Department of the Treasury   
15              National Endowment for the Arts   
16                Small Business Administration   
17            Federal Communications Commission   
18                Small Business Administration   
19            Federal Communications Commission   
20                        Department of Defense   
21              Environmental Protection Agency   
22                        Department of Justice   
23                        Department of Defense   
24     Institute of Museum and Library Services   
25        National Endowment for the Humanities   
26                  National Science Foundation   
27                      Department of Education   

                                                                                                                   business_types_description  \
0                                                                                                                                       OTHER   
1                                                                                         FOR-PROFIT ORGANIZATION (OTHER THAN SMALL BUSINESS)   
2                                                             NONPROFIT WITH 501C3 IRS STATUS (OTHER THAN AN INSTITUTION OF HIGHER EDUCATION)   
3                                                          NONPROFIT WITHOUT 501C3 IRS STATUS (OTHER THAN AN INSTITUTION OF HIGHER EDUCATION)   
4                                                                                                                                       OTHER   
5                                                             NONPROFIT WITH 501C3 IRS STATUS (OTHER THAN AN INSTITUTION OF HIGHER EDUCATION)   
6                                                             NONPROFIT WITH 501C3 IRS STATUS (OTHER THAN AN INSTITUTION OF HIGHER EDUCATION)   
7                                                                                                                                       OTHER   
8                                                                                     PUBLIC/STATE CONTROLLED INSTITUTION OF HIGHER EDUCATION   
9                                                                                     PUBLIC/STATE CONTROLLED INSTITUTION OF HIGHER EDUCATION   
10                                                                                                                                      OTHER   
11                                                                                                                                      OTHER   
12                                                            NONPROFIT WITH 501C3 IRS STATUS (OTHER THAN AN INSTITUTION OF HIGHER EDUCATION)   
13                                                            NONPROFIT WITH 501C3 IRS STATUS (OTHER THAN AN INSTITUTION OF HIGHER EDUCATION)   
14                                                            NONPROFIT WITH 501C3 IRS STATUS (OTHER THAN AN INSTITUTION OF HIGHER EDUCATION)   
15                                                            NONPROFIT WITH 501C3 IRS STATUS (OTHER THAN AN INSTITUTION OF HIGHER EDUCATION)   
16                                                                                

In [22]:
# sorting by agency and recipient
agency_recipient_sum_df = filtered_df.groupby(["awarding_agency_name", "recipient_name"])["total_outlayed_amount_for_overall_award"]\
.sum()\
.sort_values(ascending=False)\
.reset_index()

agency_recipient_sum_df

awarding_agency_name  \
0        Department of Health and Human Services   
1        Department of Health and Human Services   
2        Department of Health and Human Services   
3    Department of Housing and Urban Development   
4    Department of Housing and Urban Development   
..                                           ...   
140                Small Business Administration   
141                Small Business Administration   
142                Small Business Administration   
143                  National Science Foundation   
144                Small Business Administration   

                                                              recipient_name  \
0                                                     URBAN HEALTH PLAN, INC   
1                                                       LA CASA DE SALUD INC   
2                                                      EAST SIDE HOUSE, INC.   
3    NEW VISION COMMUNITY REDEVELOPMENT HOUSING DEVELOPMENT FUND CORPORATION   
4                                                  WEST FARMS ESTATES CO L.P   
..                                                                       ...   
140                                     BELINDA AFRICAN CARRIBEAN MARKET LLC   
141                                BRIGHT FUTURES 1-2-3 GROUP FAMILY DAYCARE   
142                                                        BUUNNI COFFEE LLC   
143                   RESEARCH FOUNDATION OF THE CITY UNIVERSITY OF NEW YORK   
144                                                              TAXI DRIVER   

     total_outlayed_amount_for_overall_award  
0                                  194957876  
1                                  183476111  
2                                   46572676  
3                                   31640391  
4                                   22627040  
..                                       ...  
140                                        0  
141                                        0  
142                                        0  
143                                        0  
144                                    -1000  

[145 rows x 3 columns]

In [23]:
# sorting by agency and recipient
agency_recipient_sum_df = filtered_df.groupby(["awarding_agency_name", "awarding_sub_agency_name", "recipient_name", "neighborhood"])["total_outlayed_amount_for_overall_award"]\
.sum()\
.sort_values(ascending=False)\
.reset_index()

agency_recipient_sum_df

awarding_agency_name  \
0        Department of Health and Human Services   
1        Department of Health and Human Services   
2        Department of Health and Human Services   
3    Department of Housing and Urban Development   
4    Department of Housing and Urban Development   
..                                           ...   
149                Small Business Administration   
150                        Department of Justice   
151      Department of Health and Human Services   
152              National Endowment for the Arts   
153                Small Business Administration   

                                          awarding_sub_agency_name  \
0                     Health Resources and Services Administration   
1                     Health Resources and Services Administration   
2                         Administration for Children and Families   
3    Assistant Secretary for Housing--Federal Housing Commissioner   
4    Assistant Secretary for Housing--Federal Housing Commissioner   
..                                                             ...   
149                                  Small Business Administration   
150                                  Offices, Boards and Divisions   
151                       Administration for Children and Families   
152                                National Endowment for the Arts   
153                                  Small Business Administration   

                                                              recipient_name  \
0                                                     URBAN HEALTH PLAN, INC   
1                                                       LA CASA DE SALUD INC   
2                                                      EAST SIDE HOUSE, INC.   
3    NEW VISION COMMUNITY REDEVELOPMENT HOUSING DEVELOPMENT FUND CORPORATION   
4                                                  WEST FARMS ESTATES CO L.P   
..                                                                       ...   
149                                                        BUUNNI COFFEE LLC   
150                                                        NEW YORK, CITY OF   
151                                            PUBLIC PREPATORY NETWORK INC.   
152                                     YOUNG URBAN CHRISTIANS & ARTISTS INC   
153                                                              TAXI DRIVER   

    neighborhood  total_outlayed_amount_for_overall_award  
0       Longwood                                194957876  
1       Longwood                                182418768  
2     Mott Haven                                 46572676  
3       Longwood                                 31640391  
4       Longwood                                 22627040  
..           ...                                      ...  
149  Hunts Point                                        0  
150      Melrose                                        0  
151     Longwood                                        0  
152      Melrose                                        0  
153      Melrose                                    -1000  

[154 rows x 5 columns]